File: acwing/算法提高课/第一章动态规划/1013.机器分配/1013.机器分配.py
N = 20
v = [0] * N
w = [0] * N
f = [[0 for _ in range(N)] for _ in range(N)]
res = [0] * N
n, m = map(int, input().split())

a = []

for i in range(n):
    a.append([0] + list(map(int, input().split())))

for i in range(0, n):
    for j in range(m, -1, -1):
        f[i][j] = f[i - 1][j]
        for k in range(1, m + 1):
            if j >= k:
                f[i][j] = max(f[i][j], f[i - 1][j - k] + a[i][k])

print(f[n - 1][m])

j = m

way = [0] * N
for i in range(n - 1, -1, -1):
    for k in range(0, j + 1):
        if f[i][j] == f[i - 1][j - k] + a[i][k]:
            way[i] = k
            j -= k
            break

for i in range(n):
    print(i+ 1,way[i])

File: acwing/算法提高课/第一章动态规划/1013.机器分配/1013.机器分配.cpp
#include <iostream>
#include <algorithm>
#include <cstring>

using namespace std;

const int N = 15, M = 20;

int f[N][M];
int a[N][M];
int way[N];

int main()
{
    int n, m;
    cin >> n >> m;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= m; j++)
        {
            cin >> a[i][j];
        }
    }
    for (int i = 1; i <= n; i++)
    {
        for (int j = 0; j <= m; j++)
        {
            for (int k = 0; k <= j; k++)
            {
                f[i][j] = max(f[i][j], f[i - 1][j - k] + a[i][k]);
            }
        }
    }
    cout << f[n][m] << endl;
    int j = m;
    for (int i = n; i > 0; i--)
    {
        for (int k = 1; k <= j; k++)
        {
            if (f[i][j] == f[i - 1][j - k] + a[i][k])
            {
                way[i] = k;
                j -= k;
                break;
            }
        }
    }
    for (int i = 1; i <= n; i++)
        cout << i << " " << way[i] << endl;
    return 0;
}